File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  const ans = [];
  
  const backtrack = (left, right, path = '') => {
    if (!left && !right) return ans.push(path);
    
    if (left) backtrack(left - 1, right, path + '(');
    if (right > left) backtrack(left, right - 1, path + ')');
  }
  
  backtrack(n, n);
  
  return ans;
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        
        @cache
        def dfs(l, r, path = ''):
            if l == 0 and r == 0:
                ans.append(path)
                return
            if l > 0:
                dfs(l - 1, r, path + '(')
            if r > l:
                dfs(l, r - 1, path + ')')
                
          
        dfs(n, n)
        return ans